code/obj/vehicle.dm

```diff
@@ -51,17 +85,87 @@ Contains:
 	Exited(atom/movable/thing, atom/newloc)
 		. = ..()
 		if(thing == src.rider)
-			src.eject_rider()
+			src.eject_rider(0, 1, 0)
 
 	proc/eject_other_stuff() // override if there's some stuff integral to the vehicle that should not be ejected
 		for(var/atom/movable/AM in src)
 			AM.set_loc(src.loc)
 
-	proc/eject_rider(var/crashed, var/selfdismount)
+	proc/eject_rider(var/crashed, var/selfdismount, var/ejectall = 1)
 		if(rider?.loc == src)
-			rider.set_loc(src.loc)
-			rider = null
-		src.eject_other_stuff()
+			src.rider.set_loc(src.loc)
+			src.overlays -= src.rider
+			src.overlays -= src.booster_image
+			handle_button_removal()
+			src.rider = null
+		if (ejectall)
+			src.eject_other_stuff()
+
+	proc/handle_button_removal()
+		if (length(src.ability_buttons))
+			for(var/obj/ability_button/B in src.ability_buttons)
+				src.rider.client?.screen -= B
+
+	proc/handle_button_addition()
+		if(!src.rider?.loc == src || !(length(src.ability_buttons)))
+			return
+		if(ishuman(src.rider))
+			var/mob/living/carbon/human/H = rider
+			H.hud?.update_ability_hotbar() //automatically adds the vehicle ability buttons
+		else if (src.rider) //fix for cannot read null.client
+			for(var/obj/ability_button/B in ability_buttons)
+				B.the_mob = src.rider
+				rider.client?.screen += B //don't have to worry about location since that should already have been handled by initialization
+
+/*			if (ishuman(rider))
+				var/mob/living/carbon/human/H = rider
+				H.hud?.update_ability_hotbar() //automatically removes the vehicle ability buttons
+			else if (length(src.ability_buttons))
+				for(var/obj/ability_button/B in src.ability_buttons)
+					rider.client.screen -= B
+*/
+
+	proc/setup_ability_buttons()
+		if (!islist(src.ability_buttons))
+			src.ability_buttons = list()
+		var/x_btt =1
+		for (var/button in src.ability_buttons_to_initialize)
+			var/obj/ability_button/NB = new button()
+			src.ability_buttons += NB
+			NB.screen_loc = "NORTH-2,[x_btt]"
+			x_btt++
+
+	relaymove(mob/user as mob, dir)
+		V_DO_MOVE(src, user, dir)
+
+	proc/do_special_on_move(mob/user as mob, dir) //empty placeholder for when we successfully have the rider relay a move
+		return
+
+/* 	relaymove(mob/user as mob, dir)
+		src.overlays = null
+		if(src.rider && user == src.rider)
```

Comment:
?

Reply:
Oh shit, forgot to remove the commented out code